Implement comprehensive rate-limiting middleware with Redis support

mclueppers opened the pull request:
Add enterprise-grade rate limiting middleware to Hockeypuck with pluggable
backend architecture supporting both memory and Redis storage for clustered
deployments.

CORE IMPLEMENTATION:
- Complete rate-limiting middleware with HTTP request interception
- Interface-based backend architecture for pluggable storage systems
- Memory backend for single-instance high-performance deployments
- Redis backend for clustered deployments with shared state
- Comprehensive Tor exit node detection and enhanced restrictions
- IP whitelisting with CIDR support for trusted networks
- Automatic ban escalation for repeat offenders

RATE LIMITING FEATURES:
- Connection-based limits (concurrent connections + connection rate)
- HTTP request rate limiting with configurable time windows
- Error rate monitoring with automatic IP banning
- Enhanced Tor exit node restrictions with escalating ban durations
- Prometheus metrics integration for monitoring and alerting
- Graceful degradation on backend failures

TOR REDIS INTEGRATION:
- Extended MetricsBackend interface with TorBackend methods
- Redis-based Tor exit node list storage for cluster-wide consistency
- Configurable useBackendStorage option for Redis/file switching
- Atomic Tor list updates with TTL-based expiration
- Local memory cache + Redis fallback for optimal performance
- Cluster-wide Tor-aware rate limiting enforcement

MIDDLEWARE CAPABILITIES:
- HTTP middleware integration with existing request pipeline
- Real-time IP extraction with proxy header support
- Context-aware rate limit decisions based on request patterns
- Automatic cleanup of stale metrics and ban records
- Configurable violation logging and ban notifications
- Zero-configuration default settings for immediate protection

CLUSTERING SUPPORT:
- Shared rate limiting state across multiple Hockeypuck instances
- Consistent ban enforcement cluster-wide via Redis backend
- Reduced memory footprint per instance in clustered deployments
- Enterprise-grade scaling with Redis replication capabilities

BACKWARD COMPATIBILITY:
- Fully backward compatible - existing deployments unchanged
- Optional rate limiting (can be disabled via configuration)
- Default file-based Tor caching preserved for single instances
- Graceful fallback mechanisms ensure service availability

FILES ADDED:
- server/ratelimit.go: Core rate limiting configuration and structures
- server/ratelimit_backend.go: Pluggable backend interface definitions
- server/ratelimit_backend_memory.go: High-performance memory backend
- server/ratelimit_backend_redis.go: Redis backend with clustering support
- server/ratelimit_core.go: Core rate limiting engine and middleware
- server/ratelimit_tor.go: Tor exit node management and restrictions
- server/ratelimit_test.go: Comprehensive test suite
- contrib/config/hockeypuck.conf.ratelimit: Example configuration

FILES MODIFIED:
- server/metrics.go: Added rate limiting Prometheus metrics
- server/server.go: Integrated rate limiting middleware
- server/settings.go: Added rate limiting configuration support
- go.mod/go.sum: Added Redis client dependency

TESTING & VERIFICATION:
- Complete unit test suite with memory and Redis backend coverage
- Integration tests for Redis connectivity and Tor storage
- End-to-end middleware testing with HTTP request simulation
- Automated verification scripts for production deployment validation

This middleware implementation provides Hockeypuck with production-ready DDoS
protection and abuse prevention while enabling horizontal scaling through
Redis-backed clustering capabilities.

Closes #387

andrewgdotcom commented on src/hockeypuck/ratelimit/cmd/ratelimit-tester/test-key.asc:
```diff
@@ -0,0 +1,29 @@
+-----BEGIN PGP PUBLIC KEY BLOCK-----
```
Is this meant to be a genuine key or nonsense? It seems to be a mishmash of genuine packets and hex gibberish ("Aa1Bb2Cc3..."). Can we use one of the existing test keys instead?

andrewgdotcom commented on src/hockeypuck/ratelimit/ratelimit.go:
```diff
@@ -0,0 +1,753 @@
+	return false, ""
+}
+
+// countRecent counts events within a time window
+func (rl *RateLimiter) countRecent(events []time.Time, window time.Duration) int {
+	cutoff := time.Now().Add(-window)
+	count := 0
+	for _, t := range events {
+		if t.After(cutoff) {
+			count++
+		}
+	}
```
This suggestion does not consider that the calling code might also be performing cleanup (it does), or that cleanup should not be performed on every invocation (it shouldn't).

andrewgdotcom commented on src/hockeypuck/server/metrics.go:
```diff
@@ -109,3 +138,19 @@ func metricsStorageNotifier(kc storage.KeyChange) error {
 }
+
+// recordRateLimitViolation records a rate limiting violation in Prometheus metrics
+func recordRateLimitViolation(reason string, isTor bool) {
```
Seems copilot got this one right though... ;-)

andrewgdotcom commented on src/hockeypuck/server/settings.go:
```diff
@@ -1,6 +1,6 @@
 /*
    Hockeypuck - OpenPGP key server
-   Copyright (C) 2012-2014  Casey Marshall
+   Copyright (C) 2012-2025 Hockeypuck Contributors
```
I've been standardising on "Casey Marshall and the Hockeypuck Contributors", on existing files at least.

andrewgdotcom commented on src/hockeypuck/server/settings.go:
```diff
@@ -244,36 +252,51 @@ func DefaultSettings() Settings {
-	tmpl, err := template.New("config").Funcs(sprig.TxtFuncMap()).Funcs(envFuncMap()).Parse(data)
-	if err != nil {
-		return nil, errors.WithStack(err)
+	// Check if data contains template syntax - if so, process as template first
```
Is there any harm just processing as a template anyway? Tiny bit of latency on startup, but surely only microseconds...?

andrewgdotcom commented on src/hockeypuck/server/settings.go:
```diff
@@ -304,3 +327,11 @@ func readEnv() map[string]string {
 }
+
+// configureDataDirPaths sets up data directory-based paths for various components
+func (s *Settings) configureDataDirPaths() {
```
This is interesting, should we do the same for other paths such as leveldb, stats etc. for consistency?

andrewgdotcom commented on Makefile:
```diff
@@ -69,10 +75,11 @@ test-postgresql:
 
 #
 # Generate targets to build Go commands.
+# Note: ratelimit-tester is in hockeypuck/ratelimit/cmd/, others are in hockeypuck/server/cmd/
```
Wondering if this is a good idea, maybe all commands should be in the same place? Not going to die on this hill though... ;-)

andrewgdotcom commented on docs/RATE_LIMITING.md:
```diff
@@ -0,0 +1,644 @@
+# Hockeypuck Rate Limiting
```
Historically we haven't had much in the way of documentation in this repo, it's mainly been in the hockeypuck.io repo. I'm wary of maintaining docs in both places, and have also been bitten a few times by people using HIPs as documentation (because they show up in search results). Need to think more long term about how we maintain docs and whether they should be consolidated. 🤔

andrewgdotcom reviewed (commented):
More nibbling round the edges of this monster :-) I haven't had the brainspace to do a proper review of the implementation itself, but just jotting down some thoughts for now.

andrewgdotcom commented on src/hockeypuck/ratelimit/ratelimit.go:
```diff
@@ -0,0 +1,753 @@
+	log.Info("rate limiter: stopping")
+
+	// Only kill and wait if Start() was called
+	if rl.started {
```
Ah, this is a much better paradigm than Peer.Idle() ... I should adopt it!

andrewgdotcom commented on docs/RATE_LIMITING.md:
```diff
@@ -0,0 +1,644 @@
+
+When enabled, IPs of configured recon partners are automatically exempted from rate limiting to ensure keyserver synchronization is not interrupted. This feature works with the HKP recon protocol configuration to identify legitimate keyserver peers.
+
+#### Keyserver Synchronization
```
Duplicate section!

mclueppers committed:
Implement comprehensive rate-limiting middleware with Redis support

Add rate limiting middleware to Hockeypuck with a pluggable backend
architecture supporting both in-memory and Redis storage for clustered
deployments.

- Interface-based backend architecture (memory + redis) under
  hockeypuck/ratelimit, registered by the server at startup.
- Connection, HTTP request, and error-rate limiting with automatic
  ban escalation for repeat offenders.
- Tor exit node detection with enhanced per-IP and global (anti-
  vandalism) restrictions; exit list fetched with a configurable
  User-Agent and cached locally.
- CIDR-based IP whitelisting and automatic exemptions for configured
  recon peers (keyserver sync).
- Prometheus metrics and rate-limit stats exposed on the stats endpoint.
- Background tasks managed via tomb.Tomb; graceful Start/Stop wired into
  the server lifecycle.

Reconciled onto current master: pksSender/sksPeer use the current
NewSender/NewPeer signatures, the stats struct uses master's layout
(RateLimit added as `any`), and dependencies are regenerated against
master's module set plus go-redis/redis/v8 and golang.org/x/net.

mclueppers committed:
Address PR #399 review feedback

- settings: always render the config as a template, and detect the
  top-level [hockeypuck] wrapper via TOML metadata instead of relying on
  a decode error. TOML silently ignores unknown top-level keys, so the
  previous "try unwrapped, fall back to wrapped" logic never ran the
  fallback and started the server with an empty Settings. Added a
  regression test. (review: TOML wrapper bug, always-template)
- settings: drop the dataDir indirection for the Tor cache path; the
  path now follows the recon ptree convention (relative to the working
  directory, overridden in config). Updated example config and docs.
- metrics: wire rate-limit metrics that were defined but never recorded.
  The ratelimit package exposes a MetricsObserver hook; the server
  registers an adapter that feeds recordRateLimitViolation /
  updateRateLimitStats on every violation, ban, and cleanup.
- tests: keep the Tor exit cache out of the package directory (no more
  orphaned tor_exit_nodes.cache); use t.TempDir()/disable caching.
- stats: use `any` instead of `interface{}` throughout the new code.
- copyright: standardise headers on "Casey Marshall and the Hockeypuck
  Contributors".
- ratelimit-tester: move under server/cmd so all commands share one
  location (removing the Makefile special-case), and ship a genuine
  test key instead of the placeholder gibberish.
- server_test: exercise a real [pks] config section.

mclueppers committed:
Address follow-up Copilot review on PR #399

- middleware: decrement the connection count synchronously with a
  background context once the request completes. The previous goroutine
  waited on the request context and then called DecrementConnections
  with that already-cancelled context, which fails on the Redis backend
  ("context canceled") and leaks the concurrent-connection count.
- middleware: sanitizeReasonForClient now matches Tor-related reasons
  before the generic "banned until" case, so active global Tor bans and
  banned Tor exits return Tor-specific messages instead of the generic
  "Access temporarily restricted". Added regression cases.
- tor: fetchTorExitList only accepts lines that parse as valid IPs,
  guarding against an upstream HTTP 200 HTML/error page polluting the
  exit list.
- tests: make the Tor fetch tests hermetic — use httptest.Server for
  status-code/User-Agent/HTML cases and a fail-fast localhost address in
  the start/stop test instead of httpbin.org; use t.TempDir() for the
  Tor stats cache file instead of a hard-coded /tmp path.
- ratelimit-tester: close each response body per iteration in
  testConnectivity instead of deferring to function return.

mclueppers committed:
Address second follow-up Copilot review on PR #399

- redis: honor the configured maxRetries on the client options.
- redis: make UpdateMetrics genuinely transactional. The WATCH callback
  previously read and wrote via b.client (plain pipelines), so the
  read/modify/write was not atomic. It now reads on the watched
  connection and writes via TxPipelined, retrying on optimistic-lock
  conflicts. GetMetrics/SetMetrics were refactored into cmdable/pipeline
  helpers to support this.
- server: only default Tor.UserAgent to the server software/version when
  it is empty, so a userAgent set in the config file takes effect.
  Updated the docs accordingly.
- ratelimit: derive the Tor exit-list fetch context from the rate
  limiter's cancellable context so Stop() aborts an in-flight fetch
  instead of blocking on the 30s timeout.
- middleware: document that check-then-track is intentionally best-effort
  (concurrent bursts from one IP may briefly overshoot limits) rather
  than a fully atomic check-and-increment across the backends.
